uswds/uswds
# [alpha] USWDS - Core: use ESM

## Summary

**USWDS Core JS now uses ES module syntax.** Scripts in uswds-core (including unit test) use ESM syntax with `.mjs` file extension. Files were previously using CommonJS syntax. Part one of migration.

## Breaking change

:warning: This is potentially a breaking change. Core file extensions have changed from `.js → .mjs` to prevent Node errors with using ESM.

## Related issue

This is the first of a few PRs for converting JS. Related to #4515 and #4987.
 
## Preview link

Preview link: n/a

## Major changes

This PR does the following:
- Adds MJS extension support to eslint and tsconfig [dc423d4eff54f7306c6aba98988a4d2cf68d4d13, ae88c3ec88dcb590da121c3c18633b2674909aed]
- Converted util scripts to ESM [8c6f646e11ba4dca8334a22edfab21b59152017d]
- Converted polyfills to ESM [1f342dc238964eaa938bc08b7c6f299b3537cf16]
- Converted core scripts to use ESM [05290fba63cfbcbe08720290755e68f8d8f8beb1]
- Updated gulp tasks [8e23c7d626839650717cbccfa22582f091aeb312]
  - Replaced `gulp-spawn-mocha` dependency because it was tied to Mocha v6, which doesn't support ESM
  - Removed duplicate typecheck task
  - Update mocha to v10
  - Updated entrypoint references to JS 
    ⚠️ Note the compile task will fail because component scripts still use CJS

## Testing and review

1. Confirm that `test.js` is only running tests on core files
2. Run `npm test`
3. All unit tests (both SASS and JS) should pass
4. Confirm that all uswds-core scripts (except `uswds-init.js`) use `import` instead of `require()`
5. Confirm that all uswds-core scripts use MJS extension (except `uswds-init.js`)

⚠️ If you encounter an error on `npm install` — that's expected. The code is compiled via a NPM post install hook. Because only core has been migrated you might see an error like the one below:

```bash
[10:06:07] Error: Can't walk dependency graph: Cannot find module './config' from '/Users/aleadem/Dev/github/uswds/packages/uswds-core/src/js/start.mjs'
    required by /Users/aleadem/Dev/github/uswds/packages/uswds-core/src/js/start.mjs
    at /Users/aleadem/Dev/github/uswds/node_modules/resolve/lib/async.js:167:35
    at load (/Users/aleadem/Dev/github/uswds/node_modules/resolve/lib/async.js:186:43)
    at onex (/Users/aleadem/Dev/github/uswds/node_modules/resolve/lib/async.js:211:17)
    at /Users/aleadem/Dev/github/uswds/node_modules/resolve/lib/async.js:24:69
    at FSReqCallback.oncomplete (node:fs:198:21)
    at FSReqCallback.callbackTrampoline (node:internal/async_hooks:130:17)
```

This is resolved in follow-up PR #5091.

## Dependency updates

| Dependency name  | Previous version | New version |
| ---------------- | :--------------: | :---------: |
| @types/node      |     16.11.19     |  18.11.18   |
| gulp-spawn-mocha |      6.0.0       |      -      |
| gulp-mocha       |        -         |    8.0.0    |
| mocha            |      6.2.3       |   10.2.0    |
| typescript       |      4.4.4       |    4.9.4    |

Dependencies updated to versions that support ESM.
